staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go

```diff
@@ -17,330 +17,238 @@ limitations under the License.
+			if err := aggregator.MergeSpecsIgnorePathConflict(merged, result.Data); err != nil {
+				return cached.NewResultErr[*spec.Swagger](fmt.Errorf("failed to build merge specs: %v", err))
+			}
 		}
-		return err
-	}
-	return nil
+		// Printing the etags list is stable because it is sorted.
+		return cached.NewResultOK(merged, fmt.Sprintf("%x", sha256.Sum256([]byte(fmt.Sprintf("%#v", etags)))))
+	}, caches)
 }
 
-// tryDeleteServiceSpecs tries delete specified specInfo from openAPISpecs map, and keeps the map intact
-// if the update fails.
-func (s *specAggregator) tryDeleteServiceSpecs(apiServiceName string) error {
-	orgSpecInfo, exists := s.openAPISpecs[apiServiceName]
+// updateServiceLocked updates the spec cache by downloading the latest
+// version of the spec.
+func (s *specAggregator) updateServiceLocked(name string) error {
+	specInfo, exists := s.specsByAPIServiceName[name]
 	if !exists {
-		return nil
-	}
-	delete(s.openAPISpecs, apiServiceName)
-	klog.V(2).Infof("Updating OpenAPI spec because %s is removed", apiServiceName)
-	defer func(start time.Time) {
-		duration := time.Since(start)
-		klog.V(2).Infof("Finished OpenAPI spec generation after %v", duration)
-
-		regenerationCounter.With(map[string]string{"apiservice": apiServiceName, "reason": "delete"})
-		regenerationDurationGauge.With(map[string]string{"reason": "delete"}).Set(duration.Seconds())
-	}(time.Now())
-	if err := s.updateOpenAPISpec(); err != nil {
-		s.openAPISpecs[apiServiceName] = orgSpecInfo
-		return err
+		return ErrAPIServiceNotFound
 	}
-	return nil
+	result := specInfo.downloader.Get()
+	filteredResult := cached.NewTransformer[*spec.Swagger](func(result cached.Result[*spec.Swagger]) cached.Result[*spec.Swagger] {
+		if result.Err != nil {
+			return result
+		}
+		return cached.NewResultOK(aggregator.FilterSpecByPathsWithoutSideEffects(result.Data, []string{"/apis/"}), result.Etag)
+	}, result)
+	specInfo.spec.Replace(filteredResult)
+	return result.Err
 }
 
-// UpdateAPIServiceSpec updates the api service's OpenAPI spec. It is thread safe.
-func (s *specAggregator) UpdateAPIServiceSpec(apiServiceName string, spec *spec.Swagger, etag string) error {
-	s.rwMutex.Lock()
-	defer s.rwMutex.Unlock()
-
-	specInfo, existingService := s.openAPISpecs[apiServiceName]
-	if !existingService {
-		return fmt.Errorf("APIService %q does not exists", apiServiceName)
-	}
-
-	// For APIServices (non-local) specs, only merge their /apis/ prefixed endpoint as it is the only paths
-	// proxy handler delegates.
-	if specInfo.apiService.Spec.Service != nil {
-		spec = aggregator.FilterSpecByPathsWithoutSideEffects(spec, []string{"/apis/"})
-	}
-
-	return s.tryUpdatingServiceSpecs(&openAPISpecInfo{
-		apiService: specInfo.apiService,
-		spec:       spec,
-		handler:    specInfo.handler,
-		etag:       etag,
-	})
+// UpdateAPIService updates the api service. It is thread safe.
+func (s *specAggregator) UpdateAPIService(apiServiceName string) error {
```

Comment:
The old name with "Spec" was better. This is not updating the apiserver. It updates the spec of the service, doesn't it?